File: server/src/main/java/com/github/bin/controller/ChatWebSocketHandler.java
package com.github.bin.controller;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.github.bin.model.Message;
import com.github.bin.service.RoomConfig;
import com.github.bin.service.RoomService;
import com.github.bin.util.JsonUtil;
import lombok.extern.slf4j.Slf4j;
import lombok.val;
import org.jetbrains.annotations.NotNull;
import org.springframework.web.socket.*;

/**
 * @author bin
 * @since 2023/08/22
 */
@Slf4j
public class ChatWebSocketHandler implements WebSocketHandler {

    @Override
    public void afterConnectionEstablished(@NotNull WebSocketSession session) {
        val roomConfig = getRoom(session);
        roomConfig.addClient(session);
    }

    @Override
    public void handleMessage(
            @NotNull WebSocketSession session, @NotNull WebSocketMessage<?> message
    ) throws Exception {
        if (message instanceof TextMessage textMessage) {
            val roomConfig = getRoom(session);
            val payload = textMessage.getPayload();
            try {
                val msg = JsonUtil.toBean(payload, Message.class);
                RoomService.handleMessage(roomConfig, session, msg);
            } catch (JsonProcessingException e) {
                session.close(new CloseStatus(4000, "消息格式错误"));
            } catch (Exception e) {
                log.warn("{} 消息处理错误: '{}'", session.getId(), payload, e);
                session.close(new CloseStatus(4000, "服务器错误:" + e.getMessage()));
            }
        }
    }

    @Override
    public void handleTransportError(@NotNull WebSocketSession session, @NotNull Throwable e) throws
            Exception {
        log.warn("{} websocket 异常", session.getId(), e);
        val roomConfig = getRoom(session);
        roomConfig.removeClient(session);
        if (session.isOpen()) {
            session.close(new CloseStatus(4000, "服务器错误:" + e.getMessage()));
        }
    }

    @Override
    public void afterConnectionClosed(@NotNull WebSocketSession session, @NotNull CloseStatus closeStatus) {
        val roomConfig = getRoom(session);
        roomConfig.removeClient(session);
    }

    @Override
    public boolean supportsPartialMessages() {
        return false;
    }

    private RoomConfig getRoom(WebSocketSession session) {
        return RoomService.get((String) session.getAttributes().get("roomId"));
    }

}

File: src/main/kotlin/com/github/bin/controller/ChatWebSocketHandler.kt
package com.github.bin.controller

import com.github.bin.model.Message
import com.github.bin.service.RoomConfig
import com.github.bin.service.RoomService
import com.github.bin.util.JsonUtil.toBean
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import org.springframework.web.socket.*
import java.io.IOException

/**
 * @author bin
 * @version 1.0.0
 * @since 2023/6/25
 */
@Component
class ChatWebSocketHandler(
        var roomService: RoomService,
) : WebSocketHandler {
    private val logger = LoggerFactory.getLogger(ChatWebSocketHandler::class.java)

    /**
     * 建立连接
     */
    override fun afterConnectionEstablished(session: WebSocketSession) {
        val roomConfig = session.getRoom()
        roomConfig += session
        logger.info("'{}' 连接 room '{}' ", session.remoteAddr, roomConfig.id)
    }

    /**
     * 接收消息
     */
    override fun handleMessage(session: WebSocketSession, message: WebSocketMessage<*>) {
        if (message is TextMessage) {
            val roomConfig = session.getRoom()
            val payload = message.payload
            try {
                val msg = payload.toBean(Message::class)
                roomService.handleMessage(roomConfig, session.id, msg)
            } catch (e: IOException) {
                logger.warn("'{}' 消息格式错误: '{}'", session.remoteAddr, payload)
                session.close(CloseStatus(4000, "消息格式错误"))
                return
            }
        }
    }

    /**
     * 发生错误
     */
    override fun handleTransportError(session: WebSocketSession, exception: Throwable) {
        val roomConfig = session.getRoom()
        roomConfig -= session
        if (session.isOpen) {
            session.close(CloseStatus(4000, "服务器错误:${exception.message}"))
        }
        logger.warn("'{}' websocket 异常", session.remoteAddr, exception)
    }

    /**
     * 关闭连接
     */
    override fun afterConnectionClosed(session: WebSocketSession, closeStatus: CloseStatus) {
        val roomConfig = session.getRoom()
        roomConfig -= session
        logger.info("'{}' 断开连接", session.remoteAddr)
    }

    /**
     * 是否支持发送部分消息
     */
    override fun supportsPartialMessages(): Boolean = false

    private fun WebSocketSession.getRoom(): RoomConfig {
        return RoomService[attributes["roomId"] as String]!!
    }

    private val WebSocketSession.remoteAddr get() = remoteAddress?.hostName ?: "unknown"
}
